shared-libs/couch-request/src/couch-request.js

```diff
@@ -0,0 +1,84 @@
+const request = require('request-promise-native');
+
+const isPlainObject = (value) => {
+  if (typeof value !== 'object' || value == null) {
+    return false;
+  }
+  const prototype = Object.getPrototypeOf(value);
+  return (prototype === null || prototype === Object.prototype || Object.getPrototypeOf(prototype) === null)
+    && !(Symbol.toStringTag in value) && !(Symbol.iterator in value);
+};
+
+const memoGlobalCouchRequestOptions = () => {
+  const cache = new Map(); // the assumption here is that this will be a fairly hot path, so caching would be valuable
```

Comment:
Because all we need is the COUCH_URL environment variable and that's immutable at runtime I think we can do even better. Let's calculate this at initialization rather than on call, and then store the `{ servername }` object in a const at the top level?

Also, let's only set `servername` if `env.PROXY_CHANGE_ORIGIN` is set.

Reply:
I'll look into it. I think one of the issues I had was that the environment isn't available at all the call sites, but I might be wrong.

Reply:
This could be an issue with testing because the env var will be set before you have a chance to mock it. There are other ways around this for testing so I think try it out and look at how you can fix the testing to keep the production code as clean as possible.

Reply:
Do you have an example?

Reply:
You can use `rewire` to set variables directly (search through the code for examples) or you can expose a function like `_init` which is called on init and also can be called by tests (this second option is a bit ugly because you're just trusting devs to ignore underscore prefixed functions).

Reply:
Thanks, I'll have a look.

Reply:
> Also, let's only set servername if env.PROXY_CHANGE_ORIGIN is set.

These are not strictly equivalent, so coupling them could block someone down the line. 

What about an env variable to enable the behaviour?